### PR TITLE
Add documentation for extension system

### PR DESCRIPTION
**Pull request by pascalwengerter**
First iteration, let's get it in quickly and then iterate as the extension system evolves

### CODE REVIEW

File: docs/extension-system/extension-types/search.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+An example of a search extension configuration can be found below:
+
+```js
+const exampleSearchExtension = {
+  id: string,
+  type: 'search',
+  scopes?: ExtensionScope[],
+  searchProvider: {
+    id: string
```
Comment (pascalwengerter) [inline docs/extension-system/extension-types/search.md:27]: We should either go for example values or a type here,  not a mix 😅 my bad!

File: docs/extension-system/_index.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+{{< toc >}}
+
+## Concepts and Building Blocks
+
+The ownCloud Web can be extended through various entry points with custom **apps** and **extensions**.
+
+### Distinction between Apps and Extensions
+
+An Application in the context of ownCloud Web is an artifact which can be distributed to and installed by an ownCloud admin.
```
Comment (dschmidt) [inline docs/extension-system/_index.md:19]: distributed to ... an admin? kinda odd to my ears

File: docs/extension-system/_index.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+It serves two main purposes:
+1. It makes the full app viewport (everything below the top bar) available to the application developer for any custom
+   application code. This includes the ability to define views with routes, navigation items for the left side bar, and more.
```
Comment (dschmidt) [inline docs/extension-system/_index.md:22]: Don't really agree with this, this is no property of an "app" - it's just that an app can register routes which are rendered in the full viewport.

File: docs/extension-system/_index.md [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+2. Through the `extensions` key in the application interface the developer can register extensions of any extension type.
+   Those extensions will become available in standardized extension points and for being queried from the extension registry
```
Comment (dschmidt) [inline docs/extension-system/_index.md:24]: ... and will be queried?

I assume you mean they will become available for being queried - but that doesn't work well with the two parts of the sentence imho

File: docs/extension-system/_index.md [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+   for custom use.
```
Comment (dschmidt) [inline docs/extension-system/_index.md:25]: what is custom use?

File: docs/extension-system/_index.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+no custom application code at all and only host extensions to be registered in the extension registry, or a combination of both.
+
+### Apps
+
+#### Technical Details
+
+To get started, define a `src/index.ts`. Below is the most basic example of its content:
+
+```ts
+// Install '@ownclouders/web-pkg' as a devDependency first (only relevant for types and autocompletion, dependency is already provided by the ownCloud Web at runtime).
```
Comment (dschmidt) [inline docs/extension-system/_index.md:37]: by ownCloud Web?

File: docs/extension-system/_index.md [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+import {
+  AppWrapperRoute,
+  ApplicationFileExtension,
+  defineWebApplication
+} from '@ownclouders/web-pkg'
+
+
+export default defineWebApplication({
+  setup({ applicationConfig }) {
+    // Here, you have access to the full injection context, meaning you can use all composable that we provide via web-pkg
```
Comment (dschmidt) [inline docs/extension-system/_index.md:47]: composables

File: docs/extension-system/_index.md [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+    // Needs to be unique within all installed extensions in any ownCloud web instance
```
Comment (dschmidt) [inline docs/extension-system/_index.md:49]: applications

File: docs/extension-system/_index.md [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    ]
+
+    // See details below
+    const routes = [
+        ...
+    ]
+
+    return {
+      appInfo: {
+        name: $gettext('Your extension'),
```
Comment (dschmidt) [inline docs/extension-system/_index.md:70]: app(lication)?

File: docs/extension-system/_index.md [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+By defining an application via `defineWebApplication` you can provide the following:
+- `appInfo` - the application metadata, which is used to make the application available via the app switcher and the app registry
+- `navItems` - the statically defined navigation items for the left side bar. Only gets rendered when more than 1 navigation item exists at runtime. 
+Additional dynamic navigation items can be registered via the extension registry.
+- `routes` - the routes to the different views of your application. May be referenced within the `navItems`. Authentication requirements can be defined per item.
+- `extensions` - the extensions to be registered in the extension registry. For more details see the `Extensions` section below.
+
+### Extensions
+
+In contrast to applications, extensions usually have a rather small scope and dedicated functionality.
```
Comment (dschmidt) [inline docs/extension-system/_index.md:91]: I find it odd to say that tbh

File: docs/extension-system/_index.md [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+
+The globally available extension registry provided by the ownCloud Web runtime can be used to both register and query extensions. All extensions
+which are being made available via an `app` get registered in the extension registry automatically. In your custom application code you can
+then query any of the available extensions by providing a `type` string and optionally a list of `scopes`. Throughout the ownCloud Web platform
+and most prominently also in the `files` app we have defined some extension points which automatically use certain extensions, see the 
+`Extension Points` section below.
+
+#### Extension Types
+
+Building an extension is limited to the extension types that are defined by the ownCloud Web extension system. See the full list of available extension types below.
```
Comment (dschmidt) [inline docs/extension-system/_index.md:103]: Why is it limited? You can build five apps that share their own extension types which are injected in one another

**Comment by kulmann** [comment]
Superseded by https://github.com/owncloud/web/pull/10400 (too dumb to work with branch protection in the fork and force-with-lease pushes)

File: docs/extension-system/_index.md [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+You're free to introduce your own extension types within your application code and use the extension registry to query the available ones. However, if you have the impression
+that an important extension type is missing, please reach out to us by opening a [GitHub issue](https://github.com/owncloud/web/issues/new/choose).
+
+#### Extension Base Configuration
+
+Any extension is required to define at least an `id` and a `type`.
+
+The `id` is supposed to be unique throughout the ownCloud Web ecosystem. In order to keep `id`s readable for humans we didn't want to enforce uniqueness through e.g. uuids. 
+Instead, we chose to use dot-formatted namespaces like e.g. `com.github.owncloud.web.files.search`. We'd like to encourage you to follow the same format for your own extensions.
```
Comment (dschmidt) [inline docs/extension-system/_index.md:122]: Maybe worth dropping a line why we require it (basically to be able to disable specific extensions via global config)

File: docs/extension-system/_index.md (hunk 3 above)
Comment (kulmann) [reply docs/extension-system/_index.md:22]: hm... I'd like to hint to what is possible, not only the most basic functionality.

File: docs/extension-system/_index.md (hunk 4 above)
Comment (kulmann) [reply docs/extension-system/_index.md:24]: Changed the paragraph to `Through the `extensions` key in the application interface you can register extensions of any extension type. Those extensions 
   are then available in standardized extension points. Additionally, they can be queried from the extension registry for 
   your own purposes.`. How about that?

File: docs/extension-system/_index.md (hunk 5 above)
Comment (kulmann) [reply docs/extension-system/_index.md:25]: `your own purposes` - whatever you want to do, go wild.

File: docs/extension-system/_index.md (hunk 6 above)
Comment (kulmann) [reply docs/extension-system/_index.md:37]: yep, typo, thanks

File: docs/extension-system/_index.md (hunk 7 above)
Comment (kulmann) [reply docs/extension-system/_index.md:47]: 👍

File: docs/extension-system/_index.md (hunk 8 above)
Comment (kulmann) [reply docs/extension-system/_index.md:49]: 👍

File: docs/extension-system/_index.md (hunk 9 above)
Comment (kulmann) [reply docs/extension-system/_index.md:70]: 👍

File: docs/extension-system/_index.md (hunk 10 above)
Comment (kulmann) [reply docs/extension-system/_index.md:91]: Do you have a different idea? 😇

File: docs/extension-system/_index.md (hunk 11 above)
Comment (kulmann) [reply docs/extension-system/_index.md:103]: It's not, you're right. Changed it to "these are the predefined ones and you can always define your own" basically.

**Pull request closed without merging**